sdk/src/main/scala/com/horizen/consensus/ConsensusDataProvider.scala

```diff
@@ -0,0 +1,104 @@
+package com.horizen.consensus
+
+import java.math.BigInteger
+
+import com.horizen.chain.SidechainBlockInfo
+import com.horizen.params.NetworkParamsUtils
+import com.horizen.storage.SidechainBlockInfoProvider
+import scorex.util.{ModifierId, ScorexLogging}
+
+import scala.annotation.tailrec
+
+trait ConsensusDataProvider {
+  this: TimeToEpochSlotConverter
+    with NetworkParamsUtils
+    with ScorexLogging {
+    val storage: SidechainBlockInfoProvider
+    val consensusDataStorage: ConsensusDataStorage
+  } =>
+
+  def getFullConsensusEpochInfoForBlock(blockId: ModifierId, blockInfo: SidechainBlockInfo): FullConsensusEpochInfo = {
+    log.debug(s"Requested FullConsensusEpochInfo for ${blockId} block id")
+
+    val previousEpochId = getPreviousConsensusEpochIdForBlock(blockId, blockInfo)
+
+    val nonceEpochInfo: NonceConsensusEpochInfo =
+      consensusDataStorage.getNonceConsensusEpochInfoOrElseUpdate(previousEpochId, calculateNonceForEpoch(previousEpochId))
+
+    val lastBlockIdInPreviousEpoch = lastBlockIdInEpochId(previousEpochId)
+    val prePreviousEpochId: ConsensusEpochId = getPreviousConsensusEpochIdForBlock(lastBlockIdInPreviousEpoch, storage.blockInfoById(lastBlockIdInPreviousEpoch))
+    val stakeEpochInfo: StakeConsensusEpochInfo =
+      consensusDataStorage
+        .getStakeConsensusEpochInfo(prePreviousEpochId)
+        .getOrElse(throw new IllegalStateException(s"Stake was not defined for epoch ${prePreviousEpochId}"))
+
+    FullConsensusEpochInfo(stakeEpochInfo, nonceEpochInfo)
+  }
+
+
+  private def calculateNonceForEpoch(epochId: ConsensusEpochId): NonceConsensusEpochInfo = {
+    val lastBlockIdInEpoch: ModifierId = lastBlockIdInEpochId(epochId)
+    val lastBlockInfoInEpoch = storage.blockInfoById(lastBlockIdInEpoch)
+
+    val nonceSource: Option[BigInteger] = foldEpoch[Option[BigInteger]](None, lastBlockIdInEpoch, lastBlockInfoInEpoch){
+      (_, blockInfo, accumulator) => (getMinimalHash(blockInfo.mainchainBlockReferenceHashes.map(_.data)), accumulator) match {
+        case (None, _) => accumulator
+        case (minimalHashOpt, None) => minimalHashOpt
+        case (Some(a), Some(b)) => Option(a.min(b))
+      }
+    }
```

Comment:
for better readability define op method first inside this methods, then call `foldEpoch`:
```
...
def op(id: ModifierId, blockInfo: SidechainBlockInfo, accumulator: Option[BigInteger]): Option[BigInteger] = {
      (getMinimalHash(blockInfo.mainchainBlockReferenceHashes.map(_.data)), accumulator) match {
        case (None, _) => accumulator
        case (minimalHashOpt, None) => minimalHashOpt
        case (Some(a), Some(b)) => Option(a.min(b))
      }
    }
    
val nonceSource: Option[BigInteger] = foldEpoch[Option[BigInteger]](None, lastBlockIdInEpoch, lastBlockInfoInEpoch)(op)
```